clippy_lints/src/collapsible_if.rs

```diff
@@ -95,14 +95,35 @@ impl CollapsibleIf {
 
     fn check_collapsible_else_if(&self, cx: &LateContext<'_>, then_span: Span, else_block: &Block<'_>) {
         if let Some(else_) = expr_block(else_block)
-            && cx.tcx.hir_attrs(else_.hir_id).is_empty()
             && !else_.span.from_expansion()
             && let ExprKind::If(else_if_cond, ..) = else_.kind
-            && !block_starts_with_significant_tokens(cx, else_block, else_, self.lint_commented_code)
         {
-            span_lint_and_then(
+            // if the inner `if` has a `#[expect(clippy::collapsible_else_if)]` attribute,
+            // we need to actually fulfill its expectation (#13365)
+            match cx.tcx.hir_attrs(else_.hir_id) {
+                [] if !block_starts_with_significant_tokens(cx, else_block, else_, self.lint_commented_code) => {},
+                // The attribute is present
+                [attr]
+                    if matches!(Level::from_attr(attr), Some((Level::Expect, _)))
+                        && let Some(metas) = attr.meta_item_list()
+                        && let Some(MetaItemInner::MetaItem(meta_item)) = metas.first()
+                        && let [tool, lint_name] = meta_item.path.segments.as_slice()
+                        && tool.ident.name == sym::clippy
+                        && matches!(
+                            lint_name.ident.name,
+                            sym::collapsible_else_if | sym::style | sym::all
+                        )
+                        // There isn't significant text apart from the attr
+                        && let span_before_attr = else_block.span.split_at(1).1.until(attr.span())
+                        && !span_contains_non_whitespace(cx, span_before_attr, self.lint_commented_code)
+                        && let span_after_attr = attr.span().between(else_.span)
+                        && !span_contains_non_whitespace(cx, span_after_attr, self.lint_commented_code) => {},
+                _ => return,
+            }
```

Comment:
This looks like the same condition as `collapsible_if` other than the lint name and can be factored out.